### PR TITLE
Fix line number cache

### LINKED ISSUE #1938: dartdoc throws exception attempting to find offset inside empty file
We've found it while running `dartdoc` on `native_widgets 1.0.0+4`.

````
Unhandled exception:
RangeError (start): Invalid value: Only valid value is 0: 1
#0      _StringBase.indexOf (dart:core/runtime/libstring_patch.dart:334:7)
#1      String.indexOf (dart:core/runtime/libstring_patch.dart:1024:18)
#2      _createLineNumbersMap (package:dartdoc/src/line_number_cache.dart:30:23)
#3      LineNumberCache.lineAndColumn.<anonymous closure> (package:dartdoc/src/line_number_cache.dart:47:21)
#4      __InternalLinkedHashMap&_HashVMBase&MapMixin&_LinkedHashMapMixin.putIfAbsent (dart:collection/runtime/libcompact_hash.dart:281:23)
#5      LineNumberCache.lineAndColumn (package:dartdoc/src/line_number_cache.dart:46:32)
#6      ModelElement.lineAndColumn (package:dartdoc/src/model.dart:3499:40)
#7      new SourceLinker.fromElement (package:dartdoc/src/source_linker.dart:88:27)
#8      ModelElement.sourceHref (package:dartdoc/src/model.dart:3222:25)
````

### PR DESCRIPTION
**Pull request by jcollins-g**
Fixes #1938.   The line number cache had some fencepost problems and this corrects them, adding tests.  Also eliminates caching of file contents that won't be used more than once, anyway.

Note: dartfmt has some ideas about formatting that seem to have changed in the last version.  See the first commit for real changes.

### CODE REVIEW

File: test/line_number_cache_test.dart [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+    test('validate empty file behavior', () {
+      File emptyFile = File(pathLib.join(_tempDir.path, 'empty_file'))
+        ..writeAsStringSync('');
+      LineNumberCache cache = LineNumberCache();
+      expect(cache.lineAndColumn(emptyFile.path, 0), equals(Tuple2(1, 0)));
+    });
+
+    test('single line without newline', () {
+      File singleLineWithoutNewline =
+          File(pathLib.join(_tempDir.path, 'single_line'))
```
Comment (pq) [inline test/line_number_cache_test.dart]: It hardly matters but I'm a little surprised to see `new`s elided here (and not elsewhere).

File: test/line_number_cache_test.dart [hunk 2]
```diff
@@ -0,0 +1,65 @@
+// Copyright (c) 2016, the Dart project authors.  Please see the AUTHORS file
```
[the remark this answers is not in the document]
Comment (jcollins-g) [reply test/line_number_cache_test.dart]: done

File: test/line_number_cache_test.dart [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+// BSD-style license that can be found in the LICENSE file.
+
+library dartdoc.cache_test;
+
+import 'dart:io';
+
+import 'package:dartdoc/src/line_number_cache.dart';
+import 'package:path/path.dart' as pathLib;
+import 'package:test/test.dart';
+import 'package:dartdoc/src/tuple.dart';
```
[the remark this answers is not in the document]
Comment (jcollins-g) [reply test/line_number_cache_test.dart]: done

File: test/line_number_cache_test.dart (hunk 1 above)
Comment (jcollins-g) [reply test/line_number_cache_test.dart]: `new` is now consistently absent.

**Comment by jcollins-g** [comment]
This results in an display error as originally written for locations that can't be found.

```
  warning: unresolved doc reference [asBroadcast]
    from dart-web_audio.ScriptProcessorNode.onAudioProcess: (file:///Users/jcollins/dart/all_sdks/2.6.0-dev.4.0/lib/web_audio/dart2js/web_audio_dart2js.dart:1:-1)
```

**Pull request closed without merging**